state/indexer/block/kv/kv.go

```diff
@@ -82,10 +87,138 @@ func (idx *BlockerIndexer) Index(bh types.EventDataNewBlockEvents) error {
 	if err := idx.indexEvents(batch, bh.Events, height); err != nil {
 		return fmt.Errorf("failed to index FinalizeBlock events: %w", err)
 	}
-
 	return batch.WriteSync()
 }
 
+func getKeys(indexer BlockerIndexer) [][]byte {
+	var keys [][]byte
+
+	itr, err := indexer.store.Iterator(nil, nil)
+	if err != nil {
+		panic(err)
+	}
+	for ; itr.Valid(); itr.Next() {
+		keys = append(keys, itr.Key())
+	}
+	return keys
+}
+
+func (idx *BlockerIndexer) Prune(retainHeight int64) (int64, int64, error) {
+	// Returns numPruned, newRetainHeight, err
+	// numPruned: the number of heights pruned or 0 in case of error. E.x. if heights {1, 3, 7} were pruned and there was no error, numPruned == 3
+	// newRetainHeight: new retain height after pruning or lastRetainHeight in case of error
+	// err: error
+
+	lastRetainHeight, err := idx.getLastRetainHeight()
+	if err != nil {
+		return 0, 0, fmt.Errorf("failed to look up last block indexer retain height: %w", err)
+	}
+	if lastRetainHeight == 0 {
+		lastRetainHeight = 1
+	}
+
+	batch := idx.store.NewBatch()
+	defer func(batch dbm.Batch) {
+		err := batch.Close()
+		if err != nil {
+			idx.log.Error(fmt.Sprintf("Error when closing pruning batch: %v", err))
+		}
+	}(batch)
+	flush := func(batch dbm.Batch) error {
+		err := batch.WriteSync()
+		if err != nil {
+			return fmt.Errorf("failed to flush pruning batch %w", err)
+		}
+		err = batch.Close()
+		if err != nil {
+			idx.log.Error(fmt.Sprintf("Error when closing pruning batch: %v", err))
+		}
+		return nil
+	}
+
+	itr, err := idx.store.Iterator(nil, nil)
+	if err != nil {
+		return 0, lastRetainHeight, err
+	}
+	deleted := 0
+	affectedHeights := make(map[int64]struct{})
+	for ; itr.Valid(); itr.Next() {
+		if keyBelongsToHeightRange(itr.Key(), lastRetainHeight, retainHeight) {
+			err := batch.Delete(itr.Key())
+			if err != nil {
+				return 0, lastRetainHeight, err
+			}
+			height := getHeightFromKey(itr.Key())
+			affectedHeights[height] = struct{}{}
+			deleted++
+		}
+		if deleted%1000 == 0 && deleted != 0 {
+			err = flush(batch)
+			if err != nil {
+				return 0, lastRetainHeight, err
+			}
+			batch = idx.store.NewBatch()
+			defer func(batch dbm.Batch) {
```

Comment:
Minor: would it be worth defining another function, in the same way `flush` is defined, to contain code in lines 161-166, and 121-125? (this would remove code duplication)